### ISSUE TITLE
Is there documentation for adding a Python source?

### ISSUE THREAD

**Issue by ELLIOTTCABLE** [Shougo/denite.nvim#403]
There is some mention in the documentation about writing new `Source` classes:

```
==============================================================================
CREATE SOURCE						*denite-create-source*

To create source, you should read default sources implementation in
rplugin/python3/denite/source/*.py.

The files are automatically loaded and denite creates new Source class object.
Source class must extend Base class in ".base".

Note: The sources must be created by Python3 language.
```

This doesn't, however, detail how to *add* the source to Denite. In attempting to write a Denite source for OCaml identifiers (replacing [Merlin's Ctrl-P-specific implementation](https://github.com/ocaml/merlin/blob/a62487c21df8e63c18da3e26a2661bdde590982f/vim/merlin/autoload/ctrlp/locate.vim)), I'm a little lost.

I saw @chemzqm's [denite-git](https://github.com/neoclide/denite-git), which only includes a `rplugin/python3/denite/source` directory, with `Source` classes in files in that directory; pursuant to that, I tried adding a `rplugin/python3/denite/source` to the Merlin vim-plugin's directory, and creating an `ocaml_identifier.py` file in it with the following content:

```python
# ============================================================================
# FILE: ocaml_identifier.py
# AUTHOR: ELLIOTTCABLE <me@ell.io>
# License: MIT license
# ============================================================================

import vim
import merlin

from .base import Base
#from denite.util import globruntime

class Source(Base):

    def __init__(self, vim):
        super().__init__(vim)
        self.vim = vim
        self.name = 'ocaml'
        self.kind = 'file'

    def on_init(self, context):
        context['__cursor_pos'] = vim.current.window.cursor

    def gather_candidates(self, context):
        try:
            pos = (context['__cursor_pos'][0], context['__cursor_pos'][1])
            merlin.vimprint(str(pos)) # FIXME
            l = merlin.command("expand-prefix", "-prefix", base,
                               "-position", merlin.fmtpos(pos))
            l = l['entries']
            l = map(lambda prop: prop['name'], l)
            l = merlin.uniq(sorted(l))
            for prop in l:
                name = prop.replace("'", "''")

            return [{
                'word': name,

                # FIXME: Copy-pasted, won't actually work. Need to figure out printing from within
                # Python, and find documentation for Merlin's output format (wtf how is there none)
                'action__path': self.vim.call(
                    'fnamemodify', self.vim.call('bufname', '%'), ':p'),
                'action__line': linenr,
                'action__col': col,
            } for name in l]

        except merlin.MerlinExc as e:
            merlin.try_print_error(e)
```

(This is probably buggy, I know nearly no Python; I'm just trying to cobble this together out of other sources I've hunted down.)

Unfortunately, Denite doesn't seem to be magically discovering this file? `:Denite ocaml_identifier` simply throws an error about that being an unknown source?

tl;dr The documentation on adding Denite sources/types/actions to your own plugins could definitely be improved! If you can help me figure out how to do this, I'm happy to give the documentation a once-over and try to improve it.

**Comment by delphinus** [comment]
I think you should do `:Denite ocaml` because your souce name is `ocaml` (there's `self.name = 'ocaml'`).

Denite loads all sources by searching `*.py` in `$VIMRUNTIME/*/rplugin/python3/denite/source`.  

* [`denite.load_sources`](https://github.com/Shougo/denite.nvim/blob/4ca86c1acd0c6b606ec975abc3f2219ba2ea5925/rplugin/python3/denite/denite.py#L200-L219)
* [`denite.util.find_rplugins`](https://github.com/Shougo/denite.nvim/blob/4ca86c1acd0c6b606ec975abc3f2219ba2ea5925/rplugin/python3/denite/util.py#L172-L185)

So you can write sources wherever path below your `&runtimepath`s.

* `~/.vim/rplugin/python3/denite/source/...`
  - your own ones
* `/path/to/denite-git/rplugin/python3/denite/source/...`
  - your installed `denite-git`

**Comment by ELLIOTTCABLE** [comment]
@delphinus Thanks for your feedback!

Turns out, unsurprisingly, the source-finding error was due to a `runtimepath` problem.

Unfortunately, now I'm stuck on figuring out how the hell `import` works in files sourced by Denite: most Vim-related python files include `import vim`; but in the context of a Denite-source, this fails.

In my code, that's fine — the `vim` instance is passed to `__init__`, which is then saved in an instance-variable; and I can replace invocations of `vim.*` with `self.vim.*`; but my code depends on [`autoload/merlin.py`](https://github.com/ocaml/merlin/blob/master/vim/merlin/autoload/merlin.py), which *does* `import vim`, and which I can't change; this eventually leads to a `ModuleNotFoundError: No module named 'vim'`, no matter what I do.

(Similarly, even getting `import merlin` itself required me to crib [some hackery](https://github.com/5t111111/denite-rails/blob/master/rplugin/python3/denite/source/rails.py#L6-L15) from `denite-rails`.)

In both of these cases, [the code that I'm trying to replicate](https://github.com/ocaml/merlin/blob/master/vim/merlin/autoload/ctrlp/locate.vim#L9-L34) gets away with a simple `import vim; import merlin` — said code is invoked by `:python3`. Reading through `:help python`, I can't see any reason why Denite would be operating under a different include-path? What am I missing about the architecture, here?

**Comment by Shougo** [comment]
`Denite ocaml` is correct.

You must not use `import vim` directly.
Please use `self.vim` instead.

> In my code, that's fine — the vim instance is passed to __init__, which is then saved in an instance-variable; and I can replace invocations of vim.* with self.vim.*; but my code depends on autoload/merlin.py, which does import vim, and which I can't change; this eventually leads to a ModuleNotFoundError: No module named 'vim', no matter what I do. Google and some further reading really haven't helped.

`autoload/merlin.py` must be called in `Vim` context.
It is not denite context.
You should create the wrapper Vim script function and called it by the source.
